### DISCUSSION TITLE
How to manually create user?

### DISCUSSION THREAD

**Discussion by hyunwoona** [nextauthjs/next-auth#2761] [Help]
### Question 💬

I am trying to seed the database in prisma, and I am manually creating a user.

```
  const testUser = await prisma.user.upsert({
    where: { email: "my.email@abcd.com" },
    update: {},
    create: {
      email: "my.email@abcd.com",
    // ...
```

However, when I try to log in using my.email@abcd.com, I cannot log in.
I understand it's not the correct way to create a user, but is there a way to populate a user record with data so I can test with it?

### How to reproduce ☕️

Create a user, with a name and your Google email
Log in with Google using that email
Log in fails

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by DinerIsmail** [comment]
I would also like to know if this is possible in order to implement an invite feature

**Comment by balazsorban44** [comment]
just use the Adapter's `createUser` method.

You might want to create an account as well, using the `linkAccount` method.

Although these steps are probably not necessary. please read the docs on extendability.

Search words: callbacks, events

**Reply by daniellin215** [reply]
I have a flow as follows: (Next-auth with Google Provider setup)

1. Existing user wants to send an email to invite fellow team member
2. After inputting the person to invite's email, invoke Google Provider's`createUser`, `linkAccount` and send the user a link to `websiteUrl/auth/signin`

Would this be possible? @balazsorban44 . I know we have email provider too with magic link, but I don't want to use it because my app currently only supports Google Provider, and I don't want same user using 2 different methods and have 2 different accounts on my backend. 

Thanks!

**Comment by daniellin215** [comment]
@hyunwoona did you end up figuring out the solution? I am using google provider, and after I call `createUser` I can see a simple user being created in the DB, however, lots of things such as `providerAccount` and it's `accessToken` and other profile related items are all missing. I get `OAuthAccountNotLinked` error and not able to call `linkAccount` here properly.

**Comment by narviii** [comment]
I'm trying to run this code to create a user, and it's not working. No error- just no changes in DB. What might be the problem?

```
const { PrismaAdapter } = require("@next-auth/prisma-adapter");
const { PrismaClient } = require('@prisma/client')

const prisma = new PrismaClient();

const adapter = PrismaAdapter(prisma)

adapter.createUser({ id: "1", name: "J Smith", email: "jsmith@example.com" })
```

**Reply by narviii** [reply]
Oh, I'm stupid. I went to ChatGPT, and our AI overlord pointed me that createUser returns a promise. So correct(and working) code is:

```
const { PrismaAdapter } = require("@next-auth/prisma-adapter");
const { PrismaClient } = require('@prisma/client')

const prisma = new PrismaClient();

const adapter = PrismaAdapter(prisma)

async function createUser() {
  await adapter.createUser({ name: "J Smith", email: "jsmith@example.com" })
  console.log("User created successfully!")
}

createUser().catch((error) => console.error(error))
```

**Reply by goodpilotfish** [reply]
Thank you for this code. Sadly, I'm getting errors importing the prisma-adapter package. The error is showing up on Node 18/19/20 and goes away if line 1 and dependent code is commented out.

Does anyone have any ideas?

```
node:internal/modules/cjs/loader:553
      throw e;
      ^

Error [ERR_PACKAGE_PATH_NOT_EXPORTED]: No "exports" main defined in /Users/<user>/playground/<folder>/buyers-app/node_modules/@auth/prisma-adapter/package.json
```

**Reply by goleary** [reply]
I tried this, but then when the adapter created user tries to sign up via the google provider I get this error message:
<img width="471" alt="image" src="https://github.com/nextauthjs/next-auth/assets/16123225/17224396-6bfb-41a8-87a7-65f40759b22c">

**Reply by SamuelR2129** [reply]
I am doing something kinda similar but not really with next-auth and auth0, using the auth0 management api to create a manager dashboard to CRUD users.

The "To confirm your identity error" seems to come about when the google/auth0 provider db is misaligned with your prisma db. 
@goleary i think since you already created the user with the adapter, they cant go through the google signup since next-auth is trying to do what you have already done via the prisma-adaptor. I would maybe recommend what i am trying to do below, I am not sure if google has a management api like auth0 though.

I havent tried the prisma-adaptor method mentioned above yet, i am praying thats my solution because i create the user with auth0 management api then also add my user to a User table in my db with prisma, now I get the "To confirm your identity" error when i try to sign in with the newly created user. It signs in okay with auth0 but its next-auth throwing the error, i think. Looking at the users i create via the standard UI signup flow, the next-auth provider creates a User entry, a Account entry (that has all the access token and scopes) and the entry in auth0, the Account entry is the missing piece of the puzzle for me.

**Reply by SamuelR2129** [reply]
My issue was i created a user in auth0 database with their management api, then added the email to my prisma db, when the new user tries to sign in, next-auth thinks 2 different users with the same email are trying to sign in. Removed the logic to update the prisma db and next-auth takes over and makes the user and account for me when they try to sign in.

**Reply by devdatkumar** [reply]
@SamuelR2129 , sorry, I didn’t fully understand what you meant. Could you please explain more clearly or make it concise (using GPT if needed)

also for the issue @goleary have mentioned. 
I'm using a custom button for OAuth to signout before signin, this way I'm able to handle the error. 

```
"use server";

import { signIn, signOut } from "@/auth";

export async function signinOAuth(formData: FormData) {
  const authProvider = formData.get("authProvider")?.toString();
  await signOut({ redirect: false });  // keep redirect: false to avoid reload/redirect, otherwise the action is disturbed. 
  await signIn(authProvider, { redirectTo: "/" }); // redirectTo is optional here.
}
```
